[cling] Decls in an unnamed namespace cannot be looked up after a unload-load cycle

### Describe the bug
A definition in an unnamed namespace cannot be redeclared by re-running the same code.

### Expected behavior
A new declaration is provided; `13` is printed twice.

### To Reproduce
```c++
// t.cxx
namespace {
  int h = 12;
}

void t() {
  printf("%d\n", ++h);
}
```
```
root [0] .x t.cxx
13
root [1] .x t.cxx
In file included from input_line_10:1:
t.cxx:6:20: error: use of undeclared identifier 'h'
  printf("%d\n", ++h);
                   ^
```

### Setup
ROOT master Ubuntu

--- a/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
+++ b/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
@@ -733,6 +733,22 @@ bool DeclUnloader::VisitRedeclarable(clang::Redeclarable<T>* R, DeclContext* DC)
   }
 
   bool DeclUnloader::VisitNamespaceDecl(NamespaceDecl* NSD) {
+    // The first declaration of an unnamed namespace, creates an implicit
+    // UsingDirectiveDecl that makes the names available in the parent DC (see
+    // `Sema::ActOnStartNamespaceDef()`).
+    // If we are reverting such first declaration, make sure we reset the
+    // anonymous namespace for the parent DeclContext so that the
+    // implicit UsingDirectiveDecl is created again when parsing the next
+    // anonymous namespace.
+    if (NSD->isAnonymousNamespace() && NSD->isFirstDecl()) {
+      auto Parent = NSD->getParent();
+      if (auto TU = dyn_cast<TranslationUnitDecl>(Parent)) {
+        TU->setAnonymousNamespace(nullptr);
+      } else if (auto NS = dyn_cast<NamespaceDecl>(Parent)) {
+        NS->setAnonymousNamespace(nullptr);
+      }
+    }
+
     // NamespaceDecl: NamedDecl, DeclContext, Redeclarable
     bool Successful = VisitDeclContext(NSD);
     Successful &= VisitRedeclarable(NSD, NSD->getDeclContext());

--- a/interpreter/cling/test/CodeUnloading/RereadFile.C
+++ b/interpreter/cling/test/CodeUnloading/RereadFile.C
@@ -23,3 +23,8 @@ macro() // CHECK: version 1
 .L macro2.h
 macro() // CHECK: 2.version 2
 //CHECK: (int) 2
+
+.x unnamedns.h
+//CHECK: 13
+.x unnamedns.h
+//CHECK-NEXT: 13

--- a/interpreter/cling/test/CodeUnloading/unnamedns.h
+++ b/interpreter/cling/test/CodeUnloading/unnamedns.h
@@ -1,0 +1,7 @@
+namespace {
+  int h = 12;
+}
+
+void unnamedns() {
+  printf("%d\n", ++h);
+}

